docs/range-proof-protocol.md

```diff
@@ -1,47 +1,32 @@
+The `range_proof` module contains an API for producing a proof that multiple integer values are within a certain range.
 
-Aggregated proof protocol
-=========================
+Aggregated range proof protocol
+===============================
 
 This is a documentation for the internal implementation of the aggregated range proof. You may find the introduction to all the pieces of the range proof protocol in the [notes](../notes/index.html) module.
 
 The aggregated range proof is a zero-knowledge proof of the following relation:
 \\[
 \operatorname{ZK-PK}\left\\{
-  v_{(j)} \in {\mathbb Z\_p}
-  : v_{(j)} \in [0, 2^n)
+  v_{(0)}, v_{(1)}, \dots, v_{(m-1)}  \in {\mathbb Z\_p}
+  : v_{(0)}, v_{(1)}, \dots, v_{(m-1)} \in [0, 2^n)
 \right\\}
 \\] 
 
-for \\(m\\) values \\(v_{(j)}\\), where \\(n\\) and \\(m\\) are both a power of \\(2\\).
+where \\(n\\) and \\(m\\) are both a power of \\(2\\).
 
-API overview
-------------
+Party and Dealer state machines
+-------------------------------
 
-The prover is composed of \\(m\\) individual parties and one dealer that coordinates messages and provides challenges.
+There are multiple ways to implement the party and dealer interactions described in the [API for the aggregated multiparty computation protocol](../aggregation/index.html#api-for-the-aggregated-multiparty-computation-protocol). One way is to model the parties and dealers as mutable objects, with a function for each step of the protocol, which mutates the party or dealer state. However, one problem with this model is that it is easy to use incorrectly - to call the functions out of order, or forget to call a function, or to call a function more than once. This deviation from the protocol could result in invalid proofs, and opens up the user to potential vulnerabilities such as replay attacks.
 
-First, parties are instantiated with their secret values and the range for the range proof. The dealer assigns a unique index \\(j\\) to each party.
+We wanted a foolproof model that would prevent users from performing any step other than the correct next step of the protocol. We treated each party and dealer as a state machine, and modeled each party and dealer state as a distinct type. We used move semantics to ensure that a previous state is consumed as it transitions to the next state. For instance, the [`PartyAwaitingValueChallenge`](../aggregation/party/struct.PartyAwaitingValueChallenge.html) type has a function [`PartyAwaitingValueChallenge::apply_challenge`](../aggregation/party/struct.PartyAwaitingValueChallenge.html#method.apply_challenge), which receives a [`ValueChallenge`](aggregation/messages/struct.ValueChallenge.html) and returns an object of the next type in the protocol ([`PartyAwaitingPolyChallenge`](../aggregation/party/struct.PartyAwaitingPolyChallenge.html)) and a commitment ([`PolyCommitment`](..aggregation/messages/struct.PolyCommitment.html)), and consumes its own state.
```

Comment:
These three paragraphs are good on their own, but I'm not sure they fit well here, since it jumps from a mathematical description above to the design tradeoffs made, then back to the math. Maybe it would be better to just describe the notation for 1 dealer / m parties here, and put this description in the `aggregation` docs above?

Reply:
👍